### PR TITLE
Don't reset LOD Bias, so user can inspect the LOD

### PR DESCRIPTION
**Pull request by sercero**
I think its better to avoid resetting the LOD because it can be annoying to have it reset when you leave the hover.

### CODE REVIEW

**Comment by paroj** [comment]
thats a valid point, however we should also adapt the UI so an accidental hover does not get you stuck in a random LOD level.

I think switching to a ListBox, where switching LOD requires a click would do

**Comment by sercero** [comment]
Done

File: ogre_mesh_viewer.py [hunk 1]
```diff
@@ -290,17 +292,27 @@ def preRenderTargetUpdate(self, evt):
...
-                ImGui.Selectable(txt, i == curr_idx)
-                if ImGui.IsItemHovered():
-                    # force this LOD level
-                    entity.setMeshLodBias(1, i, i)
+                lod_levels.append("Base Mesh" if i == 0 else "Level {}: {:.2f}".format(i, mesh.getLodLevel(i).userValue))
+
+            if ImGui.BeginListBox("##lod_levels", ImGui.ImVec2(-1, min(3, lod_count) * ImGui.GetTextLineHeightWithSpacing())):
+                for n in range(lod_count):
+                    is_selected = (self.lod_current_idx == n)
+                    if ImGui.Selectable(lod_levels[n], is_selected):
```
Comment (paroj) [inline ogre_mesh_viewer.py:307]: why not compute `txt` ad-hoc as before?

File: ogre_mesh_viewer.py [hunk 2]
```diff
@@ -290,17 +292,27 @@ def preRenderTargetUpdate(self, evt):
...
+                        self.lod_current_idx = n
+                        entity.setMeshLodBias(1, n, n)
+                    # Set the initial focus when opening the combo (scrolling + keyboard navigation focus)
+                    if is_selected:
+                        ImGui.SetItemDefaultFocus()
+                    if ImGui.IsItemHovered():
```
Comment (paroj) [inline ogre_mesh_viewer.py:313]: I dont think that we can drop this one now

File: ogre_mesh_viewer.py [hunk 3]
```diff
@@ -290,17 +292,27 @@ def preRenderTargetUpdate(self, evt):
...
+            entity.setMeshLodBias(1, self.lod_current_idx, self.lod_current_idx)  # reset LOD override
             strategy = mesh.getLodStrategy().getName()
             curr_idx = entity.getCurrentLodIndex()
             ImGui.Text("Strategy: {}".format(strategy))
+
+            lod_levels = []
             for i in range(lod_count):
-                txt = "Base Mesh" if i == 0 else "Level {}: {:.2f}".format(i, mesh.getLodLevel(i).userValue)
-                ImGui.Bullet()
-                ImGui.Selectable(txt, i == curr_idx)
```
Comment (paroj) [inline ogre_mesh_viewer.py:300]: what this was doing before is that the respective LOD in the list became active as the camera moved away. This should still work.
I think we should use `lod_current_idx=-1` for "no LOD override" and reset it to -1 if the user clicks on Base Mesh.

File: ogre_mesh_viewer.py (hunk 2 above)
Comment (sercero) [reply ogre_mesh_viewer.py:313]: Without that you can't do a quick inspection of the LOD levels like before

File: ogre_mesh_viewer.py (hunk 1 above)
Comment (sercero) [reply ogre_mesh_viewer.py:307]: I wasn't sure if it had to be precomputed, let me see...

File: ogre_mesh_viewer.py (hunk 3 above)
Comment (sercero) [reply ogre_mesh_viewer.py:300]: Oh, I didn't know about that.

Comment (sercero) [reply ogre_mesh_viewer.py:300]: What about this?

If the user moves the camera, then whatever the user has selected gets reset to what should be by distance.

Comment (paroj) [reply ogre_mesh_viewer.py:300]: on the other hand.. was not the whole point here that the user can move the camera to inspect the LOD? Like if he wants to know whether the LOD distance needs to be decreased?

Comment (sercero) [reply ogre_mesh_viewer.py:347]: I was thinking more in terms of being able to freeze the selected LOD in order to see around it.

Not in terms of moving the camera in and out.

Anyway, I think that the latest changes solve the problem.

**Comment by sercero** [comment]
> hmm.. it looks like this for me: image
> 
> and the active index gets reset while scrolling. Maybe we should go back to the bullet list as before and just keep the click to lock feature.

Yes, the idea is that when the user is scrolling then the behaviour changes to that of getting the LOD from the camera distance.

But when the user selects a LOD then that is fixed and can be inspeced regardless of distance, if it does not change.

And when the user hovers over the different LODs without selecting anything then that is just a "preview" of the LODs without fixing anything.

**Comment by paroj** [comment]
> Yes, the idea is that when the user is scrolling then the behaviour changes to that of getting the LOD from the camera distance.

ah.. I was scrolling inside the overflowing listbox..

missed how you handle the reset. I guess we need to do something more sophisticated there..

**Comment by paroj** [comment]
I guess the most elegant way would be to force a LOD level in `LodListener::prequeueEntityMeshLodChanged`.
There we can remember the natural LOD and stop forcing if that changes due to camera movement..

**Comment by sercero** [comment]
> I guess the most elegant way would be to force a LOD level in `LodListener::prequeueEntityMeshLodChanged`. There we can remember the natural LOD and stop forcing, if that changes due to camera movement..

My use case is being able to set manually the LOD without resorting to moving the camera back and forth to check the resulting LOD levels without much "effort" (that is because it might be onerous to move the camera 300 OGRE units just to check if the LOD resulted OK).

And that means being able to rotate the camera around the object at any LOD level.

Also, quickly checking the LODs by hovering over the levels is also useful in my opinion, because it gives a quick check that the LOD levels resulted OK at least on a quick glance.

To me the way it behaves now (in this pull request) includes the previous behaviour and the ones I was requesting.

The camera moving back and forth being the one who restores the regular LOD calculation to me seems a natural way to solve the issue.

**Comment by paroj** [comment]
yeah.. we are on the same page here. I was just suggesting how to detect "camera moving back and forth" in terms of LOD levels.

The current issue is, that the LOD is reset, even if you scroll inside ImGui.

**Comment by sercero** [comment]
> yeah.. we are on the same page here. I was just suggesting how to detect "camera moving back and forth" in terms of LOD levels.
> 
> The current issue is, that the LOD is reset, even if you scroll inside ImGui.

Oh, you are right haha

**Comment by sercero** [comment]
> I guess the most elegant way would be to force a LOD level in `LodListener::prequeueEntityMeshLodChanged`. There we can remember the natural LOD and stop forcing, if that changes due to camera movement..

OK, I tried using the LodListener way.

We have three variables to use: `evt.lodValue`, `evt.previousLodIndex` and `evt.newLodIndex`

The problem is that the only useful variable would be `evt.lodValue` which gives us the distance of the camera, the other variables change in a discrete manner so the user has to move the camera too much to realise that it resets the LOD behaviour.

Not only that, I would have to add a new variable to compare against `evt.lodValue` to see if it changed.

So, in my opinion is better the small annonaynce of resetting the LOD when scrolling. In which case the user would not be that concerned with it since they would be doing something different anyway.

**Comment by paroj** [comment]
I was thinking about overwriting `evt.newLodIndex` to the value we want and to stop once `newLodIndex` is different to when we started.
Clicking would then mean "keep the LOD until it is changed by the camera".

I have not tried it though. If you say it does not work, you can also check `camman.getTarget().getPosition().length()` for a threshold.

> In which case the user would not be that concerned with it since they would be doing something different anyway.

It is not as rare as you think. I was literally scrolling the LOD list when I noticed. Also, if the animations are expanded, the sidebar itself will overflow and become scrollable..

**Comment by sercero** [comment]
OK, new push.

I ended up using the `evt.lodValue` data to determine if there was a change in camera proximity to the object.

This is independent of the user being in a scrollable item of ImGui.

**Comment by paroj** [comment]
I have played around with your implementation and with my idea of using `newLodIndex` for reset.

In the end I find reset by camera motion not particularly intuitive. Therefore I went back and created a more explicit variant: #24 

what do you think?

**Comment by sercero** [comment]
> I have played around with your implementation and with my idea of using `newLodIndex` for reset.
> 
> In the end I find reset by camera motion not particularly intuitive. Therefore I went back and created a more explicit variant: #24
> 
> what do you think?

image

**Comment by paroj** [comment]
replaced by #24

**Pull request closed without merging**